### ISSUE TITLE
`jax.grad(lambda x: jnp.maximum(x, 0.0))(0.0)` returns `0.5` — deviates from standard ReLU gradient convention and from equivalent `jnp.where` formulation

### ISSUE THREAD

**Issue by JamesDLance** [jax-ml/jax#37821]
### Description

The gradient of `jnp.maximum(x, 0.0)` at `x = 0` is `0.5` in JAX. Every other major autodiff framework (PyTorch, TensorFlow) and the `jnp.where`-based equivalent return `0.0`. The `0.5` value also occurs at any input in the range `(-1.175e-38, +1.175e-38)` (float32 subnormals flush to 0 via DAZ, then the tie-breaking gradient applies), making the range of "wrong" gradients far wider than just exactly zero.

The inconsistency between two semantically identical formulations is particularly problematic:

```python
relu_max   = lambda x: jnp.maximum(x, 0.0)      # grad at 0 → 0.5
relu_where = lambda x: jnp.where(x > 0, x, 0.0) # grad at 0 → 0.0
```

## Reproducer

```python
import numpy as np
import jax
import jax.numpy as jnp

def relu_max(x):   return jnp.maximum(x, jnp.float32(0.0))
def relu_where(x): return jnp.where(x > jnp.float32(0.0), x, jnp.float32(0.0))

grad_max   = jax.grad(relu_max)
grad_where = jax.grad(relu_where)

for x in [0.0, 1e-45, -1e-45, 1e-10, -1e-10, 1.0, -1.0]:
    g_max   = float(grad_max(jnp.float32(x)))
    g_where = float(grad_where(jnp.float32(x)))
    match = "OK" if g_max == g_where else "DIFF"
    print(f"x={x:+.2e}  grad_max={g_max:.1f}  grad_where={g_where:.1f}  [{match}]")

# JVP directly
primals, tangents = jax.jvp(
    lambda x: jnp.maximum(x, jnp.float32(0.0)),
    (jnp.float32(0.0),), (jnp.float32(1.0),)
)
print(f"jvp(max(x,0), (0,), (1,)) tangent = {float(tangents)}")
```

## Expected output

```
x=+0.00e+00  grad_max=0.0  grad_where=0.0  [OK]
x=+1.00e-45  grad_max=1.0  grad_where=1.0  [OK]
x=-1.00e-45  grad_max=0.0  grad_where=0.0  [OK]
x=+1.00e-10  grad_max=1.0  grad_where=1.0  [OK]
x=-1.00e-10  grad_max=0.0  grad_where=0.0  [OK]
x=+1.00e+00  grad_max=1.0  grad_where=1.0  [OK]
x=-1.00e+00  grad_max=0.0  grad_where=0.0  [OK]
jvp(max(x,0), (0,), (1,)) tangent = 0.0
```

## Actual output

```
x=+0.00e+00  grad_max=0.5  grad_where=0.0  [DIFF]
x=+1.00e-45  grad_max=0.5  grad_where=1.0  [DIFF]   ← subnormal flushed to 0 via DAZ
x=-1.00e-45  grad_max=0.5  grad_where=0.0  [DIFF]   ← subnormal flushed to 0 via DAZ
x=+1.00e-10  grad_max=1.0  grad_where=1.0  [OK]
x=-1.00e-10  grad_max=0.0  grad_where=0.0  [OK]
x=+1.00e+00  grad_max=1.0  grad_where=1.0  [OK]
x=-1.00e+00  grad_max=0.0  grad_where=0.0  [OK]
jvp(max(x,0), (0,), (1,)) tangent = 0.5
```

## Root cause

XLA implements the JVP of `max(a, b)` as `0.5 * (1 + sign(a - b))`, producing `0.5` when `a = b`. This is the symmetric (averaged) subgradient, which is mathematically valid but:

1. Deviates from the widely-used convention `grad_relu(0) = 0`.
2. Differs from JAX's own `jnp.where`-based formulation of the same function.
3. Propagates over the entire subnormal range on CPU (float32 subnormals are flushed to 0 via XLA's DAZ mode, so every input in `(-1.175e-38, +1.175e-38)` receives gradient `0.5` instead of `0` or `1`).

## Impact

- Neural networks using `jnp.maximum(x, 0)` as a ReLU activation will compute wrong gradients at exactly-zero activations.
- The broken range extends to the full subnormal float32 range on CPU backends, making the issue more common than a measure-zero event.
- Two equivalent Python expressions produce different gradients, making it hard to refactor code without accidentally changing training dynamics.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:     0.10.1.dev20260513+64f6e229b
jaxlib:  0.10.0
numpy:   2.4.4
python:  3.12.3
OS:      Linux-6.14.0-37-generic-x86_64-with-glibc2.39
backend: cpu
```

**Comment by jakevdp** [comment]
Thanks for the report: this is working as expected; `min_p` and `max_p` specifically define their JVP rule using a `_balanced_eq` approach that returns a gradient of 0.5 at the crossover point: https://github.com/jax-ml/jax/blob/9d558c6ed1c4e762b250116fa626efa4989e75de/jax/_src/lax/lax.py#L5016-L5019

Analytically of course the gradient at the crossover is undefined, but returning NaN causes problems in practice. So we have a choice: do we return 0.0 (the limit from the left), or 1.0 (the limit from the right), or do we split the difference and return 0.5? Returning 0.5 turns out to have some nice properties in the context of gradient-based optimization, so that's what JAX returns (and I believe this convention was inherited from `autograd`, though @mattjj may be able to correct me on that).

I hope that's clear – thanks for raising the issue!
